### ISSUE TITLE
Change basic output to create multiple projects in a solution

### ISSUE THREAD

**Issue by mrlacey** [Microsoft/WindowsTemplateStudio#2252]
The intention is to change from generating a single app project that contains all code and move to an output that contains two projects in the solution.

The two projects will be

1. A UWP app project.
2. A .Net Standard library that is used by the app project.

The reasons for making this change are:
- It will allow for easier reuse of the platform agnostic code.
- It will make the testing of platform agnostic code easier as tests will not require an app context in which to run. This is beneficial for unit test scenarios, especially in a CI/CD environment.
- It will enable desirable additional functionality to be added to WTS in the future. (Including, but not limited to, Xamarin support.)

**This will be a breaking change!**
As this is a breaking change it will be released along with an update to the major version number of the tool. This will be version 3.0. While the impact on the generated projects will be small, this will require a large change to the templates used by the tool. This work will happen side-by-side with the remaining 2.x releases. - _Update to the roadmap to follow._

As a result of this change, we will be able to add the following functionality to the tool
- The generation of a unit test project to aid in testing (#299)
- Data access related features (via DB & Web) in the generated code (see issues linked from #396)
- The generation of projects in the same solution that targets Android and iOS (via Xamarin.Forms) in addition to the UWP app. (#321)
- Possible future support for more platforms.

**Possible issues and points of note**
- We will not be moving the ViewModels from the UWP project. Lots of their functionality is currently UWP specific and so would require a lot of work to make platform agnostic. It is presumed that the added complexity would not add sufficient benefit to anyone new to [UWP] development (a large part of our target audience) or anyone only looking to create a UWP app. For apps that include a lot of testable logic in their ViewModels (or CodeBehind), this will mean they get little benefit from this change as they will still need to create UWP specific test projects to exercise their code.
- It may initially look like not very much code will be moving when we make this change. This is because most of the current features are UWP specific and more platform agnostic features have been on hold while waiting for this work to be done.
- This change is likely to have an impact on how/if we add support for adding right-click support to apps not originally created by WTS as they are (presumably) less likely to have the solution structure we now expect.
- Based on the options selected, the core project may not be needed and so should not be included rather than being included but empty.
- The new project may need it's own set of localized resources. Even if we don't need to include this as part of the original functionality, we should provide [pointers to] documentation on how to add this.

**Right-Click Add**
Even though both projects form part of the application, the Right-click menu will be available from the UWP app project. This is because it may cause confusion when/if the shared project is also used with other projects/apps in the same solution.

**Action Points**
- [x] Confirm what functionality will/should move.
- [x] Review these plans once plans for Xamarin integration are established. (See #2274)
- [x] Identify the templates that need to be modified.
- [x] Investigate if we can maintain right-click update support for apps originally built with v1.x or v2.x (probably by adding the other project if it's not there.)
- [x] Update tooling relating to template creation and testing.
- [x] Update documentation related to template creation.

X-Ref #2087, #1329, & #916 

If you have concerns, questions, or strong opinions about any of the above, please comment below ASAP.

**Comment by touseefbsb** [comment]
I saw a glimpse of this is one of Build 2018 sessions , I think it was LOB uwp apps demo. where they created project with wts and output was multiple projects, even viewmodels had their own project, I was amazed bcz this is a brilliant move as it will make the seperation of concerns very modular and reuse of everything will be possible as .net standard libraries can be used anywhere.

**Comment by mrlacey** [comment]
If the core project is provided as a dependency that can be included or not as is necessary this will allow it to be added (via right-click) to any project created with v1.x or v2.x templates.
The effect of adding to an older project is that it may lead to cases where we add duplicates in the `.core` project of things that are already in the main UWP app. These will be in different namespaces and so should not cause issues. We _may_ be able to warn about this in the wizard summary but it will be up to the developer to resolve.

**Comment by sibille** [comment]
I like the idea of the dependency but see some issues with that. What if the developer renames the core project, would that result in adding a new core project each time you add a new page/feature?

An alternative would be to show a breaking change notification if we detect that project was created with 1.x or 2.x templates and there is no core project and give guidance on how to create one.

**Comment by mrlacey** [comment]
> What if the developer renames the core project

Loads of things that we do would break (or risk duplication) if the previously generated code is renamed. Why would this be different?

> An alternative would be to show a breaking change notification if we detect that project was created with 1.x or 2.x templates and there is no core project and give guidance on how to create one.

That the wizard summary shows possible issues is how we deal with this already. If the summary shows the creation of more than the developer expects, they can choose to not generate or develop in a separate folder and merge manually. Why should this be different?

Why would we give guidance on how to create a project manually when we can do it automatically? 
If we do it we can be sure that everything is created correctly and is as we expect it for future additions. If it's a manual process there are a lot of opportunities for errors or what's created to be different from what we expect. It's also creating manual work for a general operation that is the exact opposite of the goal of WTS which is to remove the need to manually do basic project infrastructure when a tool could do it for you.
Having explanatory docs would also be more documentation that needs maintenance and periodic updating.

**Comment by mrlacey** [comment]
In the single project generation, we always included the `Models` folder in the main project even if it was empty. This was to encourage using that folder/naming convention. 
In the multi-project version, we will put the "models" in the core project. As such I'm going to remove the `models` folder from the main project. The consequence of this is that we won't always have this folder in the generated solution if the selected options don't warrant creating the core project. Hopefully, as we add more data access functionality, the likelihood of generating an app without the core project--and so no 'models' folder--will be much less.

**Comment by sibille** [comment]
I think we should always create the core project, with the models folder, even if it's empty, to encourage separation of ui and core logic.

**Comment by mrlacey** [comment]
> I think we should always create the core project, with the models folder, even if it's empty, to encourage separation of ui and core logic.

Even if it's empty? What should we do to explain why there's an empty project in the solution?
Are there other placeholder folders you think should also be added?
Does this apply to all frameworks? I'd expect that someone using CodeBehind would be less concerned about separating logic.
If someone doesn't want a separate library then it'd be added for no benefit and just something in the generated solution that they'd then have to delete. I thought to add things people don't explicitly need or ask for and that they then have to manually delete was against the principles of the project.
Always creating it might seem simpler but potentially risks limiting scenarios where we need to add it to previously created apps.

**Comment by sibille** [comment]
I don't think the core project is sth a lot of people would delete. You don't need it for a quick test, that's true, but I think the enforcement of coding style is more important. For previously created projects we could still have the dependency or have a right click option that allows adding the core project.

**Comment by sibille** [comment]
I think it's hard to understand for people why this project is created if you add a chart page or image gallery (because it has sample data) but not if you add a settings page or camera page. 

Also, when we added the models folder (#94) one reason was to let people know where to put their models, I think this is still valid.

I think both aproaches are valid, let's get opinions from more people.

**Comment by mrlacey** [comment]
Some progress has been made in https://github.com/mrlacey/WindowsTemplateStudio/tree/v3.0

- Reference to the new "core" project isn't getting added due ot a race condition in the generation process causing the entry in the csproj file to be removed when trying to restore the project. (Restoration occurs before all the post actions have completed)
- Ensure dependencies are added correctly when Core project is included for functionality added with a right-click
- Make the 'Shared Core Project' hidden (only not hidden to make testing generation scenarios easier.)

**Comment by sibille** [comment]
As agreed in #2551 we'll change generate always both, the UWP and the core project

**Comment by sibille** [comment]
We've identified some more classes that could be moved to the core project:

- Json.cs, 
- TaskExtensions.cs 
- Singleton.cs

**Comment by sibille** [comment]
In #2551  we talked about adding the suffix .Uwp to Uwp project (call it App1.Uwp instead of App1)

Looking into how to do this, I found a problem: 
The $destinationdirectory$ is set by VS before starting our wizard. We can change it manually adding the Uwp suffix, but the original directory is created anyway. We could cleanup, but there  might also be other unwanted sideffects caused by changing this.

Apart, changing the uwp project name has major implications on RCA for existing projects, Devs will have to rename their projects (+ change all namesoaces) to get files added correctly

I think we should stick to the original project name + location for the Uwp project.

**Comment by mrlacey** [comment]
If this is deferred (or not done at all) then it risks creating even bigger problems when adding Xamarin support. 
I previously pushed for a smaller pain now rather than a bigger pain in the future. However, how (and when) Xamarin support will be added is still unknown so it seems reasonable to not do this now as the other issues you've found add complexity before it's needed.
We may need to explore other naming strategies as and when Xamarin support is added.

**Comment by sibille** [comment]
Verified on dev-nightly:
Templates version: 0.20.19029.2
Wizard version: 0.20.19029.1
